rfcs/0005-music-system.md

```diff
@@ -0,0 +1,155 @@
+
+As music is often streamed, there is no easy way to guarantee that an audio file begins playback immediately. Therefore there is no way to keep the foreground layers in sync with the background layer completely without using cached WAV samples. (The foreground layers are meant to provide ambience, not orchestral playback, so this is probably a non-issue.)
+
+# Alternatives
+Alternatives have been discussed extensively regarding the previous music PRs. Though much of the structure is preserved from the previous solutions, there are some key differences:
+- `Variants` are now for `tracks`, not for audio files. They are also arbitrarily defined, and are not required for the combat/non-combat switching to work properly.
+  - As it isn't reasonable to expect every `track` to have a combat `variant`, the new approach enables playing dedicated combat music in scenarios where previously regular music would have been played. This is helped by moving the combat `variants` to another `playlist` from the file specification of `tracks`.
+- `Tracks` can have multiple audio layers playing concurrently.
+  - This choice was made based on the results of the creator survey, and was not known at the time of the previous discussions.
+- `Playlists` are now "play once", not "play forever". This fixes various issues where players could get stuck listening to the same few songs when many others were also available.
+  - This change makes it much easier to have, for example, "generic human" tracks in a dedicated `playlist` while still having dedicated `tracks` in each region's own `playlist`.
+
+# Unresolved Questions
+Should `variant` support more than two `tracks` for convenience? Is there a real need?
+
+Should `playlists` have playback modes other than "shuffle"? Though we have experienced extensive issues with ordered playback, some other playback modes still might be viable.
+
+Is there a need to add weights to `playlists`?
+
+If a previously playing `track` is still being faded out when a new `playlist` is being chosen for playback, should we check if there is a valid `playlist` with this `track`? And if so, should we un-fade this `track` and resume playback?
+
+Should existing music nodes take `playlist` names instead of `tracks`?
+
+Should `playlists` keep track of which of their `tracks` have been played, and only re-shuffle when they have all been played an equal number of times? How would this interact with `variant`-based `playlist` switching?
+
+`Playlists` could have a `priority` node that allows forcibly terminating the current `playlist` in favor of another when its conditions become valid. This would simplify the condition setup for combat `playlists`, but would make it more difficult to control when these prioritized `playlists` would yield to the non-prioritized `playlists`. These are also more difficult to extend with plugins, as various plugin-made priority `playlists` could easily clash with each other.
```

Comment:
That idea feels a lot like CSS `!important`, which if used by the base implementation, is bad news for extensibility. If the base game uses priority anywhere, plugins' playlists  cannot gain priority over the base game's (short of overriding the playlist itself, i guess?). If you want to go that way, consider priority as an integer, where bigger = more important.

Reply:
Yeah, integer priorities were the idea, but they could still clash with plugins... what we would really need is float priorities so you could always slot something in between two playlists, but WeightedList doesn't support that.